src/parcsr_ls/par_relax_more_device.c

```diff
@@ -0,0 +1,520 @@
+/******************************************************************************
+ * Copyright 1998-2019 Lawrence Livermore National Security, LLC and other
+ * HYPRE Project Developers. See the top-level COPYRIGHT file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ ******************************************************************************/
+
+/******************************************************************************
+ *
+ * a few more relaxation schemes: Chebychev, FCF-Jacobi, CG  -
+ * these do not go through the CF interface (hypre_BoomerAMGRelaxIF)
+ *
+ *****************************************************************************/
+
+#include "_hypre_parcsr_ls.h"
+#include "float.h"
+
+#if defined(HYPRE_USING_CUDA)
+#include "_hypre_utilities.hpp"
+#include <thrust/for_each.h>
+#include <thrust/iterator/zip_iterator.h>
+
+HYPRE_Int  hypre_LINPACKcgtql1(HYPRE_Int *, HYPRE_Real *, HYPRE_Real *, HYPRE_Int *);
+HYPRE_Real hypre_LINPACKcgpthy(HYPRE_Real *, HYPRE_Real *);
+
+/**
+ * @brief xpay
+ *
+ * Performs
+ * y = x + a*y
+ * For vectors x,y, scalar a
+ */
+template <typename T>
+struct xpay
+{
+   typedef thrust::tuple<T &, T> Tuple;
+
+   const T scale;
+   xpay(T _scale) : scale(_scale) {}
+
+   __host__ __device__ void operator()(Tuple t) { thrust::get<0>(t) = thrust::get<1>(t) + scale * thrust::get<0>(t); }
+};
+
+/**
+ * @brief xy
+ *
+ * Performs
+ * y = x .* y
+ * For vectors x,y
+ */
+template <typename T>
+struct xy
+{
+   typedef thrust::tuple<T &, T> Tuple;
+
+   __host__ __device__ void operator()(Tuple t) { thrust::get<0>(t) = thrust::get<1>(t) * thrust::get<0>(t); }
+};
+
+/**
+ * @brief Calculates row sums and other metrics of a matrix on the device
+ * to be used for the MaxEigEstimate
+ */
+__global__ void
+hypreCUDAKernel_CSRMaxEigEstimate(HYPRE_Int      nrows,
```

Comment:
I think this kernel can be replaced with existing ones, row-sum, extract-diag and a few other thrust calls. The cost is not an issue since it's done only once. Also, not sure about necessity of using `atomic` here.

Reply:
Regrading using row-sum & extract diag: Yeah, I think that would work well, I'll think about how to implement it.
Regarding atomic: because `pos_diag`, `neg_diag` are both a single scalar used to count the number of pos/neg diag elements, all warps/blocks would be writing to it. So while not needed for in warp atomicity (as it is only ran with lane==0), it's needed for inter warp/block sync I think

Reply:
We will leave this as is, and redo the implementation after #410 is done.